### PR TITLE
[HEVC E] NumRefFrame check was corrected for Encoding Order

### PR DESCRIPTION
**Pull request by ogolikov**
[Issue]: MDP-42904
[Test]: manually

### CODE REVIEW

File: _studio/hevce_hw/h265/src/mfx_h265_encode_hw_par.cpp [hunk 1]
```diff
@@ -1450,16 +1450,31 @@ mfxStatus CheckVideoParam(MfxVideoParam& par, ENCODE_CAPS_HEVC const & caps, boo
...
-        changed ++;
-    }
-    if (par.mfx.GopRefDist > 1 && (par.mfx.NumRefFrame && par.mfx.NumRefFrame < (par.isField() ? 4 :2)) && !par.mfx.EncodedOrder)
-    {
-        par.mfx.NumRefFrame = par.isField() ? 4 :2;
-        changed ++;
+            || (par.mfx.NumRefFrame && minRefForPyramid(par.mfx.GopRefDist, par.isField()) > par.mfx.NumRefFrame)))
+    {
+		if (par.mfx.EncodedOrder &&
+		   (par.mfx.NumRefFrame > 2 && minRefForPyramid(par.mfx.GopRefDist, par.isField()) > par.mfx.NumRefFrame))
```
Comment (lakulako) [inline _studio/hevce_hw/h265/src/mfx_h265_encode_hw_par.cpp]: unneeded parenthesis for a && b && c

File: _studio/hevce_hw/h265/src/mfx_h265_encode_hw_par.cpp [hunk 2]
```diff
@@ -1450,16 +1450,31 @@ mfxStatus CheckVideoParam(MfxVideoParam& par, ENCODE_CAPS_HEVC const & caps, boo
...
-        par.m_ext.CO2.BRefType = MFX_B_REF_OFF;
-        changed ++;
-    }
-    if (par.mfx.GopRefDist > 1 && (par.mfx.NumRefFrame && par.mfx.NumRefFrame < (par.isField() ? 4 :2)) && !par.mfx.EncodedOrder)
-    {
-        par.mfx.NumRefFrame = par.isField() ? 4 :2;
-        changed ++;
+            || (par.mfx.NumRefFrame && minRefForPyramid(par.mfx.GopRefDist, par.isField()) > par.mfx.NumRefFrame)))
+    {
+		if (par.mfx.EncodedOrder &&
```
Comment (lakulako) [inline _studio/hevce_hw/h265/src/mfx_h265_encode_hw_par.cpp]: replace TABs

File: _studio/hevce_hw/h265/src/mfx_h265_encode_hw_par.cpp [hunk 3]
```diff
@@ -1450,16 +1450,31 @@ mfxStatus CheckVideoParam(MfxVideoParam& par, ENCODE_CAPS_HEVC const & caps, boo
...
+			par.bNonStandardReord = true;  // let's allow this mode in encoding order (may be special B pyr is used)
+		}
+		else
+		{
+			// disable B pyramid
+			par.m_ext.CO2.BRefType = MFX_B_REF_OFF;
+			changed++;
+		}
+    }
+    if (par.mfx.GopRefDist > 1 && (par.mfx.NumRefFrame && par.mfx.NumRefFrame < (par.isField() ? 4 :2)) )
```
Comment (lakulako) [inline _studio/hevce_hw/h265/src/mfx_h265_encode_hw_par.cpp]: unneeded parenthesis

File: _studio/hevce_hw/h265/src/mfx_h265_encode_hw_par.cpp [hunk 4]
```diff
@@ -1450,16 +1450,31 @@ mfxStatus CheckVideoParam(MfxVideoParam& par, ENCODE_CAPS_HEVC const & caps, boo
...
+    {
+		if (par.mfx.EncodedOrder && par.isField() && par.mfx.NumRefFrame > 1 && par.mfx.NumRefFrame < 4)
```
Comment (lakulako) [inline _studio/hevce_hw/h265/src/mfx_h265_encode_hw_par.cpp]: par.mfx.NumRefFrame < 4 is true from the condition above

File: _studio/hevce_hw/h265/include/mfx_h265_encode_hw_utils.h [hunk 5]
```diff
@@ -800,6 +800,7 @@ class MfxVideoParam : public mfxVideoParam, public TemporalLayers
     bool   bMBQPInput;
     bool   RAPIntra;
     bool   bFieldReord;
+    bool   bNonStandardReord; // Possible in encoding order only. No NumRefFrames limitation.
```
Comment (lakulako) [inline _studio/hevce_hw/h265/include/mfx_h265_encode_hw_utils.h]: Why it is not the same as EncodingOrder? In EncodingOrder the order parameters verification should be minimal, I think.

**Comment by onabiull** [comment]
Will reissue this one with corrections

**Pull request closed without merging**